### DISCUSSION TITLE
Strudel + Hydra = 🔥

### DISCUSSION THREAD

**Discussion by felixroos** [tidalcycles/strudel#565] [Ideas]
Let's talk about ways to combine Hydra and Strudel.. came up [here](https://discord.com/channels/779427371270275082/937365093082079272/1104477312768028682).

Already it's possible to use [strudel in the hydra repl](https://hydra.ojack.xyz/?code=bGV0JTIwc3RydWRlbCUyMCUzRCUyMGF3YWl0JTIwaW1wb3J0KCdodHRwcyUzQSUyRiUyRmNkbi5za3lwYWNrLmRldiUyRiU0MHN0cnVkZWwuY3ljbGVzJTJGY29yZScpJTNCJTIwJTBBbGV0JTIwJTdCbWluaSU3RCUyMCUzRCUyMGF3YWl0JTIwaW1wb3J0KCdodHRwcyUzQSUyRiUyRmNkbi5za3lwYWNrLmRldiUyRiU0MHN0cnVkZWwuY3ljbGVzJTJGbWluaScpJTNCJTBBc3RydWRlbC5QYXR0ZXJuLnByb3RvdHlwZSU1Qid2YWx1ZSclNUQlMjAlM0QlMjBmdW5jdGlvbiUyMCgpJTIwJTdCcmV0dXJuJTIwdGhpcy5xdWVyeShuZXclMjBzdHJ1ZGVsLlN0YXRlKG5ldyUyMHN0cnVkZWwuVGltZVNwYW4od2luZG93LnRpbWUlMkN3aW5kb3cudGltZSkpKSU1QjAlNUQudmFsdWUlN0QlM0IlMEElMEFvc2MoKCklMjAlM0QlM0UlMjAxMCUyMColMjBtaW5pKCcwJTIwMSUyMCU1QjIlMjA0JTVEJykuc2xvdygyKS52YWx1ZSgpKS5vdXQoKSUwQSUwQQ%3D%3D) and [the other way around](https://strudel.tidalcycles.org/?yyC0H8uPXNrL).
Throwing some questions in:

- Should hydra be integrated into the Strudel REPL by default? Are there naming collisions?
- How could hydra params be controlled by patterns?
- How could visual / audio sync work?

edit: links:

- this project also uses strudel + hydra https://github.com/folz/hiasynth

**Comment by yaxu** [comment]
An earlier experiment with @ojack: https://www.youtube.com/watch?v=Jvs7Q4cdLy4

**Comment by ojack** [comment]
Writing down some quick thoughts here!

**1. using strudel logic to control hydra parameters:** 
In that earlier experiment with yaxu, I made a version of hydra that would type-check an argument passed to hydra, and if it was a strudel pattern, evaluate it as such. I also tried out evaluating any string as a mini-notation pattern, so you could do `osc("2 3 4").out()` . I considered making this the default behavior in hydra. With the latter, I got feedback that some people are using strings in hydra in different ways and so it might not be a good idea...however I still wan't to dig into this more, and possibly make it an opt-in feature!!

**2. syncronizing strudel audio and hydra visuals**
I can't remember exactly what we did in the video above, but there was something a bit hacky that we did in order to use shared patterns in both hydra and strudel. When a function is passed as a hydra parameter, hydra evaluates that function on every frame update, and uses the result as a parameter. There is usually (but doesn't have to be) a concept of time used in the function so that it is independent of frame rate. For an external function to be usable in hydra, it should be able to return a number that represents the value of that thing at the moment that the function is evaluated. My understanding is that strudel programs events to happen in the future. So in order to syncronize the things, we had to do some weird time subtraction. sorry if this isn't so specific, trying to remember things. 

**3. using strudel in hydra editor**
I would love to have a straightforward way to use strudel notation within hydra, as well as to program strudel audio directly in the hydra editor (i haven't tried this yet!!) in a way that the clocks of both are synchronized. In general with hydra, I am moving away from pre-bundling lots of libraries, but working on making it as easy as possible to load external libraries. This includes having a UI listing extensions people have made to hydra libraries + libraries compatible with hydra, and examples directly accessible from the editor. I am imagining something like:

``` await loadScript("path/to/strudel-hydra") ```

that would enable a lot of the features listed above and possibly enable the default string parsing. If someone wanted something with more finite control, they could load different libraries individually as done in the hydra repl example above. 

**4. using hydra in strudel editor**
happy to help with whatever makes this easier :)

**Reply by yaxu** [reply]
1 - ah nice! In the main strudel deployment, @felixroos has [some transpiling stuff](https://github.com/tidalcycles/strudel/tree/main/packages/transpiler) so double quoted strings are automatically parsed as mininotation. To specify an actual string, you can still use single quotes. Without this or similar, you have to put `mini()` around every mininotation string within strudel code.

2. hmm I don't remember clearly either.. but yep I guess you need to get them in phase somehow, and preferably have the same cycle count. Back then I know that strudel always ran at 1 cycle per second. it can run a different tempos now which probably complicates things further..
3. That'd be great! I guess there'd need to be some hook so that strudel could replace hydra's clock source, or vice versa

**Reply by felixroos** [reply]
> In the main strudel deployment, @felixroos has [some transpiling stuff](https://github.com/tidalcycles/strudel/tree/main/packages/transpiler)

because the transpiler only works on code strings, this will not work when the pattern functions are used in regular JS land, as string arguments will just be string arguments. But there is another trick for that: [setStringParser](https://github.com/tidalcycles/strudel/blob/a03541295627ee7f1a6df545a61dd9ab7b68d421/packages/core/pattern.mjs#L22) can be used to define a custom handler for strings passed to `reify`, which is called on every argument of every pattern function. So calling `setStringParser(mini)` will have a similar effect than the transpiler, but any string will be interpreted as mini notation, so you cannot use regular strings (as you can use with single quotes). Still, the transpiler is useful to find out where the mini notation is located, so the visual highlighting works in the editor. For hydra, something like `setStringParser` could maybe work for opt-in logic

**4 - using hydra in strudel editor (but also kind of related to all of the above)**
It would be great to have some way of highlighting events that are not ending up in the strudel scheduler (like when they are passed to hydra). In this [example](https://strudel.tidalcycles.org/?JtBHMePreay2), the mini pattern passed to the osc function is not highlighted, because it is not part of the main strudel pattern (thus not queried by the strudel highlighting animation loop). So there needs to be a way to collect all "external" patterns and stack them to the main pattern just for the highlighting loop. If hydra had something like `setStringParser`, something like that could do the trick:

```js
setStringParser((str) => {
  const pat = mini(str);
  addToHighlighter(pat);
  return pat.value();
}); 
```

though that would only work for mini patterns, maybe hydra could have a more generalized hook like `addArgumentHandler` or whatever... which nicely circles back to:

> a version of hydra that would type-check an argument passed to hydra [...] possibly make it an opt-in feature

```js
addArgumentHandler((arg, skip) => {
  if (typeof arg === "string") {
    const pat = mini(str);
    addToHighlighter(pat);
    return pat.value();
  }
  if (typeof arg === "function") {
    const pat = arg();
    if (!isPattern(pat)) {
      return skip();
    }
    addToHighlighter(pat);
    return pat.value();
  }
  return skip();
});
```

the skip function is just a way to tell "i don't want to handle that argument".
this idea might still have flaws.. but it looks ok when I blink a few times at 60fps

**Comment by ojack** [comment]
Example from @froos of how to play strudel sound within hydra editor:::

https://pastebin.com/LnJbzF4P


I am currently working on a UI for directly loading external libraries and extensions into the hydra editor, so will definitely add the strudel one!!!

<img width="500" alt="Screenshot 2023-07-18 at 17 14 24" src="https://github.com/tidalcycles/strudel/assets/4909769/b768fce6-c24d-4266-9017-4fa9f606ff8c">
<img width="500" alt="Screenshot 2023-07-18 at 17 14 16" src="https://github.com/tidalcycles/strudel/assets/4909769/c4fa4e8d-c182-4e10-a986-657050ee0e61">

**Comment by jarmitage** [comment]
Dropping this in here too cc @atfornes https://github.com/atfornes/Hydra-strudel-extension

**Comment by felixroos** [comment]
strudel hydra integration is now also available since https://github.com/tidalcycles/strudel/pull/759 see https://strudel.cc/learn/hydra
